v2/spiffeid/id.go

```diff
@@ -2,77 +2,159 @@ package spiffeid
 
 import (
 	"net/url"
+	"strings"
+
+	"github.com/zeebo/errs"
 )
 
+var idErr = errs.Class("spiffeid")
+
 // ID is a SPIFFE ID
 type ID struct {
+	td   TrustDomain
+	path string
 }
 
 // New creates a new ID using the trust domain (e.g. example.org) and path
 // segments. An error is returned if the trust domain is not valid (see
 // TrustDomainFromString).
+// Warning: Percent encoded characters on path segments are
+// not decoded, instead this function will percent encode symbols
+// when the ID is converted to a string.
 func New(trustDomain string, segments ...string) (ID, error) {
-	panic("not implemented")
+	td, err := TrustDomainFromString(trustDomain)
+	if err != nil {
+		return ID{}, err
+	}
+
+	path := strings.Join(segments, "/")
+	if len(path) > 0 {
```

Comment:
We should only add the `"/"` prefix if the path doesn't already have it (e.g. New("example.org", "/foo", "bar))

Reply:
Wouldn't  this make the New func a bit inconsistent? Why the first segment is different than the others? Why we would add a slash before "/bar" in this case New("example.org", "foo", "/bar) then?

Reply:
That is true. In that case, for all of these `segments ...string` based functions, we should probably be using `path.Join` instead. It will `Clean()` the result after doing the join, essentially normalizing the path. We can then add the `/` prefix (if necessary) to ensure that the path is consistent. If you want a path with empty segments or what-not, then use `spiffeid.FromString("spiffe://example.org/some//weird/////path")`.

Reply:
So, spiffeid.Must("domain.test", "", "path", "element") would produce a path as "/path/element", right?

Reply:
I think that is fine as long we document the behavior. People continue to have an escape hatch for "non-traditional" paths via `FromString` but our behavior for other functions is to produce a clean, normalized path.